Add Text.measure

Gadgetoid opened the pull request:
Returns the size of displayed text in pixels.

Useful for getting the width/height of text for aligning it on an unknown screen size.

For example, to center text horizontally and vertically:

```
tw, th = text.measure("Hello world!")
text.move((board.DISPLAY.width - tw) / 2, (board.DISPLAY.height - th) / 2)
text.text("Hello world!")
```

Note: this code assumes the font is 8px by 16px.

deshipu commented on stage.py:
```diff
@@ -589,6 +589,17 @@ def text(self, text, hightlight=False):
+                cols += 1
+            if self.column >= self.width or c == '\n':
+                rows += 1
+        return cols * 8, rows * 16
```
The font used in Stage is 8x8.

deshipu commented on stage.py:
```diff
@@ -589,6 +589,17 @@ def text(self, text, hightlight=False):
+        cols = 0
+        for c in text:
+            if ord(c) >= 32:
+                cols += 1
```
This will keep incrementing ``cols`` for every line, so if you have 3 lines of text with 10 characters each, you will end up with ``cols=30``. What you want is to reset ``cols`` to ``0`` on each line as ``text()`` does, and keep a maximum of all lines.

deshipu commented on stage.py:
```diff
@@ -589,6 +589,17 @@ def text(self, text, hightlight=False):
                 if self.row >= self.height:
                     self.row = 0
 
+    def measure(self, text, hightlight=False):
```
I wonder if it would make more sense to simply return the dimensions in the ``text()`` call, since you are going to call it anyways, and it doesn't matter in what order you do it.

Gadgetoid replied:
This is a good idea- I think it did cross my mind (I could be inventing memories here), but I didn't want to make breaking changes.

deshipu replied:
It won't break any existing code — in Python you can call a method and ignore its return value, and there won't be any errors.

Gadgetoid commented:
Okay I started to work this into the `text` method and ended up running into a few bugs caused by the function stepping through text a char at a time.

I have rewritten the `text` method with the help of a new method `split` which preprocesses the text string to:

1. Split on `\n`
2. Split the result if it's wider than the text buffer

This yields individual strings which can be drawn line-by-line with no further consideration for wrap/line breaks.

I realize I might be treading on eggshells - and endangering back-compat with low-mem devices - making this more complicated, but hopefully the use of a generator will keep memory usage low.

## `Text.text` returns the size of displayed text in pixels.

Useful for getting the width/height of text for aligning it on an unknown screen size.

For example, to center text horizontally and vertically:

```
tw, th = text.text("Hello world!")
text.move((board.DISPLAY.width - tw) / 2, (board.DISPLAY.height - th) / 2)
```

## Add "align_center" to `Text.text`

Aligns text to the center of the container by padding lines with spaces.

Eg:

```python
text = stage.Text(24, 5)
tw, th = text.text("""Hello
Hello World
Long Hello World
Really really long line which will wrap""", align_center=True)
text.move((board.DISPLAY.width - tw) / 2, (board.DISPLAY.height - th) / 2)
```
will produce:

![stage-text-center](https://user-images.githubusercontent.com/1746967/135470293-6f284f66-2d5d-4ed0-ae1a-fdb101f35cee.jpg)

## Add `Text.split`

Splits text into chunks on '\n' and subdivides those on buffer width.

yields individual lines of split, wrapped text.

deshipu commented:
I was mostly thinking about something along the lines of:

```python

    def text(self, text, hightlight=False):
        """Display text starting at the current cursor location."""
        longest = 0
        tallest = 0
        for c in text:
            if ord(c) >= 32:
                self.char(self.column, self.row, c, hightlight)
                self.column += 1
            if self.column >= self.width or c == '\n':
                longest = max(longest, self.column)
                self.column = 0
                self.row += 1
                if self.row >= self.height:
                    tallest = max(tallest, self.row)
                    self.row = 0
        longest = max(longest, self.column)
        tallest = max(tallest, self.row)
        return longest * 8, (tallest + (1 if self.column > 0 else 0)) * 8
```

The centering feature is nice, but it should probably be a separate pull request — we also need to decide where we want to stop, as typesetting is a huge topic and it could easily take a lot of additional code (just look at TeX).

deshipu commented:
If you don't mind, I'm going to go with just adding the measurements to ``text()`` and close this for now.

deshipu commented:
Implemented by https://github.com/python-ugame/circuitpython-stage/commit/03caa2bf36ccc37a4c336715a7aa447b337e950a

Gadgetoid commented:
I don't mind at all! Sorry I was dragging my heels, too many spinning plates!
